test/performance/odp_ipsecfwd.c

```diff
@@ -0,0 +1,1307 @@
+	return true;
+}
+
+static inline odp_ipsec_sa_t *get_in_sa(odp_packet_t pkt)
+{
+	odph_esphdr_t esp;
+	uint32_t spi;
+
+	if (!odp_packet_has_ipsec(pkt))
+		return NULL;
+
+	if (odp_packet_copy_to_mem(pkt, odp_packet_l4_offset(pkt), ODPH_ESPHDR_LEN, &esp) < 0)
+		return NULL;
+
+	spi = odp_be_to_cpu_32(esp.spi);
+
+	return spi <= UINT16_MAX ? spi_to_sa_map[DIR_IN][spi] : NULL;
+}
+
+static inline odp_bool_t process_ipsec_in(odp_packet_t pkt, const odp_ipsec_sa_t *sa)
+{
+	odp_ipsec_in_param_t param;
+
+	memset(&param, 0, sizeof(param));
+	param.num_sa = 1;
+	param.sa = sa;
+	/* IPsec in/out need to be identified somehow, so use user_ptr for this. */
+	odp_packet_user_ptr_set(pkt, NULL);
+
+	return odp_ipsec_in_enq(&pkt, 1, &param) == 1;
+}
+
+static inline odp_ipsec_sa_t *get_out_sa(odp_packet_t pkt)
+{
+	odph_udphdr_t udp;
+	uint16_t dst_port;
+
+	if (!odp_packet_has_udp(pkt))
+		return NULL;
+
+	if (odp_packet_copy_to_mem(pkt, odp_packet_l4_offset(pkt), ODPH_UDPHDR_LEN, &udp) < 0)
+		return NULL;
+
+	dst_port = odp_be_to_cpu_16(udp.dst_port);
+
+	return dst_port ? spi_to_sa_map[DIR_OUT][dst_port] : NULL;
+}
+
+static inline odp_bool_t process_ipsec_out(odp_packet_t pkt, const odp_ipsec_sa_t *sa)
+{
+	odp_ipsec_out_param_t param;
+
+	memset(&param, 0, sizeof(param));
+	param.num_sa = 1;
+	param.sa = sa;
+	/* IPsec in/out need to be identified somehow, so use user_ptr for this. */
+	odp_packet_user_ptr_set(pkt, sa);
+
+	return odp_ipsec_out_enq(&pkt, 1, &param) == 1;
+}
+
+static inline const pktio_t *lookup_and_apply(odp_packet_t pkt, odph_table_t fwd_tbl,
+					      odp_bool_t is_eth_mod)
+{
+	const uint32_t l3_off = odp_packet_l3_offset(pkt);
+	odph_ipv4hdr_t ipv4;
+	uint32_t ip, l2_off;
+	fwd_entry_t *fwd;
+	odph_ethhdr_t eth;
+
+	if (odp_packet_copy_to_mem(pkt, l3_off, ODPH_IPV4HDR_LEN, &ipv4) < 0)
+		return NULL;
+
+	ip = odp_be_to_cpu_32(ipv4.dst_addr);
+
+	if (odph_iplookup_table_get_value(fwd_tbl, &ip, &fwd, 0U) < 0 || fwd == NULL)
+		return NULL;
+
+	if (is_eth_mod) {
+		if (l3_off > ODPH_ETHHDR_LEN) {
+			if (odp_packet_pull_head(pkt, l3_off - ODPH_ETHHDR_LEN) == NULL)
+				return NULL;
+		} else {
+			if (odp_packet_push_head(pkt, ODPH_ETHHDR_LEN - l3_off) == NULL)
+				return NULL;
+		}
+
+		eth.type = odp_cpu_to_be_16(ODPH_ETHTYPE_IPV4);
+		l2_off = 0U;
+	} else {
+		l2_off = odp_packet_l2_offset(pkt);
+
+		if (odp_packet_copy_to_mem(pkt, l2_off, ODPH_ETHHDR_LEN, &eth) < 0)
+			return NULL;
+	}
+
+	eth.dst = fwd->dst_mac;
+	eth.src = fwd->pktio->src_mac;
+
+	if (odp_packet_copy_from_mem(pkt, l2_off, ODPH_ETHHDR_LEN, &eth) < 0)
+		return NULL;
+
+	return fwd->pktio;
+}
+
+static inline odp_bool_t forward_packet(odp_packet_t pkt, odph_table_t fwd_tbl, stats_t *stats,
+					int worker_id, odp_bool_t is_eth_mod)
+{
+	const pktio_t *pktio = lookup_and_apply(pkt, fwd_tbl, is_eth_mod);
+
+	if (pktio == NULL) {
+		++stats->discards;
+		return false;
+	}
+
+	if (odp_unlikely(pktio->send_fn(pktio, worker_id % pktio->num_tx_qs, pkt) < 0)) {
```

Comment:
This leaks packets if odp_pktout_send() returns zero. The problem is that odp_pktout_queue() and odp_queue_enq() have a bit different return values but they are used directly in send_fn functions.

The queue index parameter so send_fn is a bit nasty since it is not used at all by enqueue() but probably is still calculated even in that case.

There is also the issue that with multiple output queues packet order gets lost. It is not clear if it is a bug or a feature. Preserving packet order with multiple output queues would be a bit complicated but could be done e.g. by selecting the queue based on traffic flow rather than thread ID (but that may not be enough since I think direct mode output even with single queue does not preserve packet order the same way as queue mode output).

Reply:
Fixed in v4. Depending on configured `mode`, the output will be ordered (one output queue) or parallel with several queues. This enables testing a few different queue setups. Queue index is always calculated but the overhead isn't particularly large.

Reply:
ok, this can be tweaked later if there is a need.